### PR TITLE
Kernel/Net: Misc work

### PR DESCRIPTION
**Pull request by logkos**
This PR:
1. Removes obsolete diagnostic skip from EFH.h
2. merges EtherType.h and EFH.h
3. Introduces SO_SNDBUF and SO_RCVBUF for getsocketopt

### CODE REVIEW

**Review by sdomi — COMMENTED** [review]
all of those changes feel very feng-shui to me. furthermore, `UNIX` directory will only contain two files - is there a point here?

This is largely analogous to your previous contribution that shuffled a bunch of IP stuff around (a part of which we're currently reverting)

File: Kernel/Net/Ethernet.h [hunk 1]
```diff
@@ -8,8 +8,17 @@
 
 #include <AK/Endian.h>
 #include <AK/MACAddress.h>
+#include <AK/Types.h>
 
-#pragma GCC diagnostic ignored "-Warray-bounds"
```
Comment (sdomi) [inline Kernel/Net/Ethernet.h:12]: is there a reason for removing this line? As in, did this throw a warning before and it doesn't anymore..?

File: Kernel/Net/EtherType.h
Comment (sdomi) [inline Kernel/Net/EtherType.h:1]: ... is there a point? both of those files are small, yes, but they're only partially related to each other.

File: Kernel/Net/UNIX/LocalSocket.h
Comment (sdomi) [inline Kernel/Net/UNIX/LocalSocket.h:1]: the `UNIX` directory will only contain two files - is there some greater point with this change?

Furthermore, @kleinesfilmroellchen has a WiP branch that greatly refactors this and other socket-related files. It may be wise to wait for that to land before changing the paths.

File: Kernel/Net/EtherType.h
Comment (kleinesfilmroellchen) [inline Kernel/Net/EtherType.h:1]: This first commit is very much feng shui programming, which is against our contribution guidelines. There is no tangible benefit; in fact, everyone wanting to access the ether type enum now has to include the whole ethernet frame header.

File: Kernel/Net/Ethernet.h (hunk 1 above)
Comment (kleinesfilmroellchen) [inline Kernel/Net/Ethernet.h:12]: Removing this diagnostic will probably break at least the GCC build. If it is actually not required anymore, please make a separate commit removing it, since in general we shouldn't ignore diagnostics unless the compiler is very clearly wrong about something (and that usually doesn't happen unless we're doing whacky bleeding-edge-features things).

File: Kernel/Net/LocalSocket.cpp
Comment (kleinesfilmroellchen) [inline Kernel/Net/LocalSocket.cpp:1]: The second commit is making a useful change, but the implementation is both stylistically and logically wrong and should probably be rewritten entirely.

File: Kernel/Net/Ethernet.h (hunk 1 above)
Comment (logkos) [reply Kernel/Net/Ethernet.h:12]: I think It was resolved after adding packed attribute

File: Kernel/Net/LocalSocket.cpp [hunk 2]
```diff
@@ -399,14 +399,23 @@ ErrorOr<void> LocalSocket::getsockopt(OpenFileDescription& description, int leve
...
     MutexLocker locker(mutex());
 
+    int buffer_size;
     socklen_t size;
     TRY(copy_from_user(&size, value_size.unsafe_userspace_ptr()));
 
     switch (option) {
     case SO_SNDBUF:
-        return ENOTSUP;
+        buffer_size = sizeof(*m_for_server.leak_ptr());
```
Comment (kleinesfilmroellchen) [inline Kernel/Net/LocalSocket.cpp:408]: This line is wrong on multiple levels, but most importantly: the size of the DoubleBuffer structure does not represent the buffer size, since the buffer is dynamically allocated. Please read up on the relevant APIs to find out how to actually get the buffer size.

File: Kernel/Net/LocalSocket.cpp [hunk 3]
```diff
@@ -399,14 +399,23 @@ ErrorOr<void> LocalSocket::getsockopt(OpenFileDescription& description, int leve
...
+        TRY(copy_to_user(static_ptr_cast<int*>(value), &buffer_size));
```
Comment (kleinesfilmroellchen) [inline Kernel/Net/LocalSocket.cpp:409]: static_ptr_cast is only for smart pointers (and in fact, only for RefPtr currently). Additionally, you are not checking that the user allocated us enough space for a size_t.

File: Kernel/Net/LocalSocket.cpp [hunk 4]
```diff
@@ -399,14 +399,23 @@ ErrorOr<void> LocalSocket::getsockopt(OpenFileDescription& description, int leve
 
     MutexLocker locker(mutex());
 
+    int buffer_size;
```
Comment (kleinesfilmroellchen) [inline Kernel/Net/LocalSocket.cpp:402]: Buffer sizes are not 32-bit integers. They are size_t, whatever that means on the current architecture (64 bits or more, and unsigned).

File: Kernel/Net/LocalSocket.cpp [hunk 5]
```diff
@@ -399,14 +399,23 @@ ErrorOr<void> LocalSocket::getsockopt(OpenFileDescription& description, int leve
...
+        TRY(copy_to_user(static_ptr_cast<int*>(value), &buffer_size));
+        size = sizeof(buffer_size);
+        TRY(copy_to_user(value_size, &size));
+        return {};
     case SO_RCVBUF:
-        return ENOTSUP;
+        buffer_size = sizeof(*m_for_client.leak_ptr());
+        TRY(copy_to_user(static_ptr_cast<int*>(value), &buffer_size));
+        size = sizeof(buffer_size);
+        TRY(copy_to_user(value_size, &size));
```
Comment (kleinesfilmroellchen) [inline Kernel/Net/LocalSocket.cpp:417]: Same issues here.

File: Kernel/Net/EtherType.h
Comment (logkos) [reply Kernel/Net/EtherType.h:1]: Having a common dump field for Ethernet attributes seemed more viable for the sake of cleanliness and future expansions

File: Kernel/CMakeLists.txt [hunk 6]
```diff
@@ -276,7 +276,6 @@ set(KERNEL_SOURCES
     Net/Realtek/RTL8168NetworkAdapter.cpp
     Net/VirtIO/VirtIONetworkAdapter.cpp
     Net/IPv4/Socket.cpp
-    Net/LocalSocket.cpp
```
Comment (kleinesfilmroellchen) [inline Kernel/CMakeLists.txt]: More feng shui programming. We don't need a separate subdirectory for two files. The IP subdirectory (which you renamed recently to IPv4, and which we will rename back) will soon contain close to ten files, which is a level at which subdirectories actually make sense.

File: Kernel/Net/UNIX/LocalSocket.h
Comment (logkos) [reply Kernel/Net/UNIX/LocalSocket.h:1]: This might have a point in far future when stack grows to include multiple protocols, reverting tho

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]
While the second commit contains a change that would, in its goals, be useful, its implementation is totally wrong and the other two commits are entirely unnecessary. I'd suggest you discard the work done here and have another go at implementing the buffer size support for local socket `getsockopt`.

File: Kernel/Net/Ethernet.h (hunk 1 above)
Comment (sdomi) [reply Kernel/Net/Ethernet.h:12]: That's incorrect. The pragma was introduced with ac215ca601393dc111643b119835ea39e6c9bd23, while the packed struct was already there. 

It's a very old commit, so it's not unlikely that it's indeed not needed anymore - but removing it blindly is something I'm not happy about.

Comment (kleinesfilmroellchen) [reply Kernel/Net/Ethernet.h:12]: Alright then, please make a separate commit. Such changes are always welcome.

**Comment by kleinesfilmroellchen** [comment]
The first commit should still be dropped, which means the other two commits are totally unrelated and should live in separate PRs.
